File: Ventas.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Scanner;

public class Ventas {

    private List<List<Double>> matrizVentas;
    private String[] meses = {
        "Enero", "Febrero", "Marzo", "Abril", "Mayo", "Junio",
        "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre"
    };
    private List<String> departamentos;

    public Ventas() {
        this.departamentos = new ArrayList<>();
        this.departamentos.add("Ropa");
        this.departamentos.add("Deportes");
        this.departamentos.add("Juguetería");

        this.matrizVentas = new ArrayList<>();
        for (int i = 0; i < 12; i++) {
            this.matrizVentas.add(new ArrayList<>());
        }
        generarVentasAleatorias();
    }

    public void generarVentasAleatorias() {
        Random random = new Random();
        for (int i = 0; i < 12; i++) {
            this.matrizVentas.get(i).clear();
            for (int j = 0; j < departamentos.size(); j++) {
                double monto = 1000 + (20000 - 1000) * random.nextDouble();
                this.matrizVentas.get(i).add(Math.round(monto * 100.0) / 100.0);
            }
        }
    }

    public void agregarDepartamento(String nombreDepto) {
        departamentos.add(nombreDepto);
        for (int i = 0; i < 12; i++) {
            matrizVentas.get(i).add(0.0);
        }
        System.out.println("\n ¡Departamento '" + nombreDepto + "' agregado a la lista!");
    }

    public void insertarVenta(int mesNum, int deptoNum, double monto) {
        if (mesNum >= 1 && mesNum <= 12 && deptoNum >= 1 && deptoNum <= departamentos.size()) {
            matrizVentas.get(mesNum - 1).set(deptoNum - 1, monto);
            System.out.printf("\n Venta de $%.2f registrada en %s (%s).\n", 
                monto, departamentos.get(deptoNum - 1), meses[mesNum - 1]);
        } else {
            System.out.println("\n Error: Número de mes o departamento inválido.");
        }
    }

    public void insertarVentaSilenciosa(int mesNum, int deptoNum, double monto) {
        if (mesNum >= 1 && mesNum <= 12 && deptoNum >= 1 && deptoNum <= departamentos.size()) {
            matrizVentas.get(mesNum - 1).set(deptoNum - 1, monto);
        }
    }

    public double buscarVenta(int mesNum, int deptoNum) {
        if (mesNum >= 1 && mesNum <= 12 && deptoNum >= 1 && deptoNum <= departamentos.size()) {
            double monto = matrizVentas.get(mesNum - 1).get(deptoNum - 1);
            System.out.printf("\n Venta en %s (%s): $%.2f\n", 
                departamentos.get(deptoNum - 1), meses[mesNum - 1], monto);
            return monto;
        } else {
            System.out.println("\n Error: Número de mes o departamento inválido.");
            return -1;
        }
    }

    public void eliminarVenta(int mesNum, int deptoNum) {
        if (mesNum >= 1 && mesNum <= 12 && deptoNum >= 1 && deptoNum <= departamentos.size()) {
            matrizVentas.get(mesNum - 1).set(deptoNum - 1, 0.0);
            System.out.printf("\n️ Venta eliminada (reiniciada a $0.00) en %s (%s).\n", 
                departamentos.get(deptoNum - 1), meses[mesNum - 1]);
        } else {
            System.out.println("\n Error: Número de mes o departamento inválido.");
        }
    }

    public void mostrarTabla() {
        System.out.println("\n------------------- TABLA DE VENTAS -------------------");
        System.out.printf("%-12s | ", "Mes");
        for (String dept : departamentos) {
            System.out.printf("%-12s | ", dept);
        }
        System.out.println("\n--------------------------------------------------------");

        double[] totalesAnuales = new double[departamentos.size()];

        for (int i = 0; i < 12; i++) {
            System.out.printf("%-12s | ", meses[i]);
            for (int j = 0; j < departamentos.size(); j++) {
                double venta = matrizVentas.get(i).get(j);
                totalesAnuales[j] += venta;
                System.out.printf("$%-11.2f | ", venta);
            }
            System.out.println();
        }

        System.out.println("--------------------------------------------------------");
        System.out.printf("%-12s | ", "TOTAL ANUAL");
        for (int j = 0; j < departamentos.size(); j++) {
            System.out.printf("$%-11.2f | ", totalesAnuales[j]);
        }
        System.out.println("\n--------------------------------------------------------");
    }

    public void mostrarDepartamentos() {
        System.out.println("\nDepartamentos disponibles:");
        for (int i = 0; i < departamentos.size(); i++) {
            System.out.println((i + 1) + ". " + departamentos.get(i));
        }
    }

    public int getCantidadDepartamentos() {
        return departamentos.size();
    }

    public String getNombreMes(int mesNum) {
        return meses[mesNum - 1];
    }

    public static void main(String[] args) {
        // Mide el tiempo de inicialización del programa y carga del primer menú
        long inicioArranque = System.currentTimeMillis();

        Ventas sistema = new Ventas();
        Scanner scanner = new Scanner(System.in);
        int opcion = 0;

        long finArranque = System.currentTimeMillis();
        long msArranque = finArranque - inicioArranque;
        System.out.printf("Tiempo de carga e inicialización del programa: %.3f segundos (%d ms)\n", 
                          msArranque / 1000.0, msArranque);

        while (opcion != 7) {
            System.out.println("\n====================================");
            System.out.println("    SISTEMA DE CONTROL DE VENTAS    ");
            System.out.println("====================================");
            System.out.println("1. Ver tabla de ventas");
            System.out.println("2. Insertar / Modificar venta");
            System.out.println("3. Buscar venta");
            System.out.println("4. Eliminar venta");
            System.out.println("5. Agregar nuevo departamento");
            System.out.println("6. Regenerar datos aleatorios");
            System.out.println("7. Salir");
            System.out.print("Selecciona una opción (1-7): ");

            if (scanner.hasNextInt()) {
                opcion = scanner.nextInt();
                scanner.nextLine(); // Limpieza del buffer
            } else {
                scanner.nextLine();
                System.out.println("\n Opción no válida.");
                continue;
            }

            // --- INICIO DE MEDICIÓN DE TIEMPO DE LA OPCIÓN ---
            long inicioTiempo = System.currentTimeMillis();

            if (opcion == 1) {
                sistema.mostrarTabla();
            } else if (opcion == 2) {
                System.out.print("Ingrese número de mes (1-12): ");
                int m = scanner.nextInt();
                sistema.mostrarDepartamentos();
                System.out.print("Ingrese número de departamento: ");
                int d = scanner.nextInt();
                System.out.print("Ingrese el monto: ");
                double monto = scanner.nextDouble();
                scanner.nextLine();
                sistema.insertarVenta(m, d, monto);
            } else if (opcion == 3) {
                System.out.print("Ingrese número de mes (1-12): ");
                int m = scanner.nextInt();
                sistema.mostrarDepartamentos();
                System.out.print("Ingrese número de departamento: ");
                int d = scanner.nextInt();
                scanner.nextLine();
                sistema.buscarVenta(m, d);
            } else if (opcion == 4) {
                System.out.print("Ingrese número de mes (1-12): ");
                int m = scanner.nextInt();
                sistema.mostrarDepartamentos();
                System.out.print("Ingrese número de departamento: ");
                int d = scanner.nextInt();
                scanner.nextLine();
                sistema.eliminarVenta(m, d);
            } else if (opcion == 5) {
                System.out.print("\nIngrese el nombre del nuevo departamento: ");
                String nombreDepto = scanner.nextLine();
                sistema.agregarDepartamento(nombreDepto);
                
                System.out.println("\n¿Qué desea hacer con los datos de venta de '" + nombreDepto + "'?");
                System.out.println("1. Mantener todas las ventas en $0.00");
                System.out.println("2. Ingresar los montos mes a mes manualmente");
                System.out.print("Seleccione una opción (1 o 2): ");
                
                int opcionLlenado = 1;
                if (scanner.hasNextInt()) {
                    opcionLlenado = scanner.nextInt();
                    scanner.nextLine();
                } else {
                    scanner.nextLine();
                }

                if (opcionLlenado == 2) {
                    int nuevoDeptoId = sistema.getCantidadDepartamentos();
                    System.out.println("\n--- Ingresando ventas para: " + nombreDepto + " ---");
                    
                    for (int m = 1; m <= 12; m++) {
                        System.out.print("Monto para " + sistema.getNombreMes(m) + ": $");
                        double monto = scanner.nextDouble();
                        sistema.insertarVentaSilenciosa(m, nuevoDeptoId, monto);
                    }
                    scanner.nextLine();
                    System.out.println("\n ¡Todas las ventas de '" + nombreDepto + "' han sido registradas exitosamente!");
                } else {
                    System.out.println("Las ventas de '" + nombreDepto + "' se han inicializado en $0.00.");
                }

            } else if (opcion == 6) {
                sistema.generarVentasAleatorias();
                System.out.println("\n Nuevas ventas aleatorias generadas.");
            } else if (opcion == 7) {
                System.out.println("\n¡Hasta luego!");
            }

            // --- FIN DE MEDICIÓN DE TIEMPO Y CÁLCULO ---
            long finTiempo = System.currentTimeMillis();
            long milisegundos = finTiempo - inicioTiempo;
            double segundos = milisegundos / 1000.0;

            if (opcion >= 1 && opcion <= 7) {
                System.out.printf("\n Tiempo transcurrido para procesar la opción %d: %.3f segundos (%d ms)\n", 
                                  opcion, segundos, milisegundos);
            }
        }
        scanner.close();
    }
}

File: Ventas.py
import random
import time


class Ventas:

    def __init__(self):
        self.meses = [
            "Enero",
            "Febrero",
            "Marzo",
            "Abril",
            "Mayo",
            "Junio",
            "Julio",
            "Agosto",
            "Septiembre",
            "Octubre",
            "Noviembre",
            "Diciembre",
        ]
        self.departamentos = ["Ropa", "Deportes", "Jugueteria"]
        self.matriz_ventas = [[] for _ in range(12)]
        self.generar_ventas_aleatorias()

    def generar_ventas_aleatorias(self):
        for i in range(12):
            self.matriz_ventas[i].clear()
            for _ in range(len(self.departamentos)):
                monto = random.uniform(1000, 20000)
                self.matriz_ventas[i].append(round(monto, 2))

    def agregar_departamento(self, nombre_depto):
        self.departamentos.append(nombre_depto)
        for i in range(12):
            self.matriz_ventas[i].append(0.0)
        print(f"\nDepartamento '{nombre_depto}' agregado a la lista.")

    def insertar_venta(self, mes_num, depto_num, monto):
        if (
            1 <= mes_num <= 12
            and 1 <= depto_num <= len(self.departamentos)
        ):
            self.matriz_ventas[mes_num - 1][depto_num - 1] = monto
            print(
                f"\nVenta de ${monto:.2f} registrada en"
                f" {self.departamentos[depto_num - 1]} ({self.meses[mes_num - 1]})."
            )
        else:
            print("\nError: Numero de mes o departamento invalido.")

    def insertar_venta_silenciosa(self, mes_num, depto_num, monto):
        if (
            1 <= mes_num <= 12
            and 1 <= depto_num <= len(self.departamentos)
        ):
            self.matriz_ventas[mes_num - 1][depto_num - 1] = monto

    def buscar_venta(self, mes_num, depto_num):
        if (
            1 <= mes_num <= 12
            and 1 <= depto_num <= len(self.departamentos)
        ):
            monto = self.matriz_ventas[mes_num - 1][depto_num - 1]
            print(
                f"\nVenta en {self.departamentos[depto_num - 1]}"
                f" ({self.meses[mes_num - 1]}): ${monto:.2f}"
            )
            return monto
        else:
            print("\nError: Numero de mes o departamento invalido.")
            return -1

    def eliminar_venta(self, mes_num, depto_num):
        if (
            1 <= mes_num <= 12
            and 1 <= depto_num <= len(self.departamentos)
        ):
            self.matriz_ventas[mes_num - 1][depto_num - 1] = 0.0
            print(
                f"\nVenta eliminada (reiniciada a $0.00) en"
                f" {self.departamentos[depto_num - 1]} ({self.meses[mes_num - 1]})."
            )
        else:
            print("\nError: Numero de mes o departamento invalido.")

    def mostrar_tabla(self):
        print("\n------------------- TABLA DE VENTAS -------------------")
        header = f"{'Mes':<12} | " + "".join(
            [f"{dept:<12} | " for dept in self.departamentos]
        )
        print(header)
        print("-" * len(header))

        totales_anuales = [0.0] * len(self.departamentos)

        for i in range(12):
            fila = f"{self.meses[i]:<12} | "
            for j in range(len(self.departamentos)):
                venta = self.matriz_ventas[i][j]
                totales_anuales[j] += venta
                fila += f"${venta:<11.2f} | "
            print(fila)

        print("-" * len(header))
        fila_total = f"{'TOTAL ANUAL':<12} | " + "".join(
            [f"${tot:<11.2f} | " for tot in totales_anuales]
        )
        print(fila_total)
        print("-" * len(header))

    def mostrar_departamentos(self):
        print("\nDepartamentos disponibles:")
        for i, dept in enumerate(self.departamentos, start=1):
            print(f"{i}. {dept}")

    def get_cantidad_departamentos(self):
        return len(self.departamentos)

    def get_nombre_mes(self, mes_num):
        return self.meses[mes_num - 1]


def main():
    inicio_arranque = time.time()

    sistema = Ventas()

    fin_arranque = time.time()
    segundos_arranque = fin_arranque - inicio_arranque
    ms_arranque = int(segundos_arranque * 1000)
    print(
        f"Tiempo de carga e inicializacion del programa:"
        f" {segundos_arranque:.3f} segundos ({ms_arranque} ms)"
    )

    opcion = 0

    while opcion != 7:
        print("\n====================================")
        print("    SISTEMA DE CONTROL DE VENTAS    ")
        print("====================================")
        print("1. Ver tabla de ventas")
        print("2. Insertar / Modificar venta")
        print("3. Buscar venta")
        print("4. Eliminar venta")
        print("5. Agregar nuevo departamento")
        print("6. Regenerar datos aleatorios")
        print("7. Salir")

        try:
            opcion = int(input("Selecciona una opcion (1-7): "))
        except ValueError:
            print("\nError: Opcion no valida. Por favor ingresa un numero.")
            continue

        inicio_tiempo = time.time()

        if opcion == 1:
            sistema.mostrar_tabla()

        elif opcion == 2:
            try:
                m = int(input("Ingrese numero de mes (1-12): "))
                sistema.mostrar_departamentos()
                d = int(input("Ingrese numero de departamento: "))
                monto = float(input("Ingrese el monto: "))
                sistema.insertar_venta(m, d, monto)
            except ValueError:
                print("\nError: Entrada numerica invalida.")

        elif opcion == 3:
            try:
                m = int(input("Ingrese numero de mes (1-12): "))
                sistema.mostrar_departamentos()
                d = int(input("Ingrese numero de departamento: "))
                sistema.buscar_venta(m, d)
            except ValueError:
                print("\nError: Entrada numerica invalida.")

        elif opcion == 4:
            try:
                m = int(input("Ingrese numero de mes (1-12): "))
                sistema.mostrar_departamentos()
                d = int(input("Ingrese numero de departamento: "))
                sistema.eliminar_venta(m, d)
            except ValueError:
                print("\nError: Entrada numerica invalida.")

        elif opcion == 5:
            nombre_depto = input("\nIngrese el nombre del nuevo departamento: ")
            sistema.agregar_departamento(nombre_depto)

            print(
                f"\nQue desea hacer con los datos de venta de '{nombre_depto}'?"
            )
            print("1. Mantener todas las ventas en $0.00")
            print("2. Ingresar los montos mes a mes manualmente")

            try:
                opcion_llenado = int(input("Seleccione una opcion (1 o 2): "))
            except ValueError:
                opcion_llenado = 1

            if opcion_llenado == 2:
                nuevo_depto_id = sistema.get_cantidad_departamentos()
                print(f"\n--- Ingresando ventas para: {nombre_depto} ---")

                for m in range(1, 13):
                    while True:
                        try:
                            monto = float(
                                input(
                                    f"Monto para {sistema.get_nombre_mes(m)}: $"
                                )
                            )
                            sistema.insertar_venta_silenciosa(
                                m, nuevo_depto_id, monto
                            )
                            break
                        except ValueError:
                            print(
                                "Error: Ingrese un valor numerico valido para"
                                " el monto."
                            )

                print(
                    f"\nTodas las ventas de '{nombre_depto}' han sido"
                    " registradas exitosamente."
                )
            else:
                print(
                    f"Las ventas de '{nombre_depto}' se han inicializado en"
                    " $0.00."
                )

        elif opcion == 6:
            sistema.generar_ventas_aleatorias()
            print("\nNuevas ventas aleatorias generadas.")

        elif opcion == 7:
            print("\nHasta luego!")

        fin_tiempo = time.time()
        segundos = fin_tiempo - inicio_tiempo
        milisegundos = int(segundos * 1000)

        if 1 <= opcion <= 7:
            print(
                f"\nTiempo transcurrido para procesar la opcion {opcion}:"
                f" {segundos:.3f} segundos ({milisegundos} ms)"
            )


if __name__ == "__main__":
    main()
